src/eventstore/WriteThroughFileCheckpoint.cs

```diff
@@ -1,95 +1,46 @@
-   As such we should not add a Corsham Science copyright file header */
-#pragma warning disable CA1001, CA1060, CA2101
+﻿// <copyright file="WriteThroughFileCheckpoint.cs" company="Corsham Science">
+// Copyright (c) Corsham Science. All rights reserved.
+// </copyright>
+#pragma warning disable CA1001
 
 namespace CorshamScience.MessageDispatch.EventStore
 {
-    using System;
     using System.IO;
-    using System.Runtime.InteropServices;
-    using System.Threading;
-    using Microsoft.Win32.SafeHandles;
 
     /// <summary>
     /// Writes a checkpoint to a file pulled from event store.
     /// </summary>
     internal class WriteThroughFileCheckpoint
     {
-        private readonly FileStream _stream;
-        private readonly bool _cached;
-        private readonly BinaryWriter _writer;
+        private readonly object _streamLock = new ();
+        private readonly FileStream _fileStream;
         private readonly BinaryReader _reader;
-        private readonly MemoryStream _memStream;
-        private readonly byte[] _buffer;
-
-        private long _last;
-        private long _lastFlushed;
+        private readonly BinaryWriter _writer;
+        private long? _lastWritten;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="WriteThroughFileCheckpoint"/> class.
         /// </summary>
-        /// <param name="filename">The file to write a checkpoint to.</param>
-        /// <param name="name">The name of the checkpoint to write.</param>
-        /// <param name="cached">Indicates if the checkpoint has been cached.</param>
+        /// <param name="filePath">The file to write a checkpoint to.</param>
         /// <param name="initValue">The initial value to write.</param>
-        public WriteThroughFileCheckpoint(string filename, string name, bool cached, long initValue = 0)
+        public WriteThroughFileCheckpoint(string filePath, long initValue = 0L)
         {
-            Name = name;
-            _cached = cached;
-            _buffer = new byte[4096];
-            _memStream = new MemoryStream(_buffer);
-
-            var handle = Filenative.CreateFile(
-                filename,
-                (uint)FileAccess.ReadWrite,
-                (uint)FileShare.ReadWrite,
-                IntPtr.Zero,
-                (uint)FileMode.OpenOrCreate,
-                Filenative.FileFlagNoBuffering | (int)FileOptions.WriteThrough,
-                IntPtr.Zero);
+            var alreadyExisted = File.Exists(filePath);
 
-            _stream = new FileStream(handle, FileAccess.ReadWrite, 4096);
-            var exists = _stream.Length == 4096;
-            _stream.SetLength(4096);
-            _reader = new BinaryReader(_stream);
-            _writer = new BinaryWriter(_memStream);
+            _fileStream = new FileStream(filePath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.ReadWrite);
 
-            if (!exists)
+            if (_fileStream.Length != 8)
```

Comment:
Probably safest to use sizeof(long) here I guess.

Reply:
Done.